k2/csrc/fsa_algo.h

```diff
@@ -659,6 +660,41 @@ void InvertHost(FsaOrVec &src, Ragged<int32_t> &src_aux_labels, FsaOrVec *dest,
 FsaOrVec RemoveEpsilonSelfLoops(FsaOrVec &src,
                                 Array1<int32_t> *arc_map = nullptr);
 
+
+/*
+  Replace, in `index`, `labels symbol_range_begin <= label < symbol_range_begin+src.Dim0()`
+  with the Fsa indexed `label - symbol_range_begin` in `src`, identifying the
+  source and destination states of the arc in `src` with the initial and
+  final states in `src[label - symbol_range_begin]`. Arcs with labels outside
+  this range are just copied. Labels on final-arcs in `src` (which will be -1)
+  would be set to 0(epsion) in the result.
+
+    @param [in] src    FsaVec containing individual Fsas that we'll be inserting
+                       into the result.
+    @param [in] index  Fsa or FsaVec (2 or 3 axes) that dictates the overall
+                       structure of the result (the result will have the same
+                       number of axes as`index`.
+
+    @param [in] symbol_range_begin  Beginning of the range (interval) of symbols
+                                    that are to  be replaced with Fsas.
+                                    Symbols numbered `symbol_range_begin <= i < src.Dim0() + symbol_range_begin`
+                                    will be replaced with the Fsa in `src[i - symbol_range_begin]`
+
+    @param [out,optional] arc_map_src  If not nullptr, will be set to a new
+                                       array that maps from arc-indexes in the
```

Comment:
Actually (my fault) this comment is not quite accurate, or at least not specific.
Because this arc_map will be used in autograd, we need to clarify that it is only set for the 
initial arc corresponding each label in `src` that is replaced.  I assume this is an initial epsilon
in the output that leads into the start-state of the element of `index`?  (Otherwise I'm wondering
how we deal with the situation where there are arcs into the start-state of the FSA in `index`).

Reply:
In my current implementation, the following `index` and `src`
|index|src|
|--|--|
|![image](https://user-images.githubusercontent.com/11765074/123365549-0bdaa580-d5a9-11eb-9bc6-748d6964e0d7.png)|![image](https://user-images.githubusercontent.com/11765074/123365572-172dd100-d5a9-11eb-8bf8-205a12ca306d.png)|

replace with `symbol_range_begin=2` is:

![image](https://user-images.githubusercontent.com/11765074/123365598-23b22980-d5a9-11eb-933f-9ebacdbdacec.png)

and `arc_map_src = [-1, -1, 0, 1, 2, 3, -1]`, mapping arcs to src FsaVec, all of the arcs inherit from `index` would be -1.
`arc_map_index = [0, 1, -1, -1, -1, -1, 2]`, mapping arcs to index Fsa, arcs be replaced(epsilon arcs that leading to the start-state of corresponding fsa in `src`) and arcs with label out-of-range will have values, the others will be -1.

As for the situation where there are arcs into the state-state of the FSA in `index`,  current implementation can not handle it appropriate, will fix it.